File: javascript/packages/linter/src/rules/actionview-no-redundant-local-assigns.ts
import { ParserRule } from "../types.js"
import { PrismVisitor, Visitor } from "@herb-tools/core"

import { isPrismNodeType, isRubyParameterNode, locationFromByteOffset } from "@herb-tools/core"
import { isPartialFile } from "./file-utils.js"

import type { DocumentNode, ERBStrictLocalsNode, ParseResult, ParserOptions, PrismNodes, RubyParameterNode } from "@herb-tools/core"
import type { FullRuleConfig, LintContext, UnboundLintOffense } from "../types.js"

const KEYWORD_KIND = "keyword"
const KEYWORD_REST_KIND = "keyword_rest"
const LOCAL_ASSIGNS = "local_assigns"

const VALUE_LOOKUPS = ["[]", "fetch", "dig"]
const PRESENCE_LOOKUPS = ["key?", "has_key?", "include?", "member?"]

interface Lookup {
  name: string
  method: string
  presence: boolean
  extraArguments: boolean
  startOffset: number
  length: number
}

class StrictLocalsCollector extends Visitor {
  hasDeclaration = false
  hasKeywordRest = false

  readonly locals = new Map<string, RubyParameterNode>()

  visitERBStrictLocalsNode(node: ERBStrictLocalsNode): void {
    this.hasDeclaration = true

    for (const local of node.locals) {
      if (!isRubyParameterNode(local)) continue

      if (local.kind === KEYWORD_REST_KIND) {
        this.hasKeywordRest = true
        continue
      }

      if (local.kind !== KEYWORD_KIND) continue

      const name = local.name?.value

      if (name) this.locals.set(name, local)
    }
  }
}

class LocalAssignsLookupCollector extends PrismVisitor {
  readonly lookups: Lookup[] = []

  override visitCallNode(node: PrismNodes.CallNode): void {
    const lookup = this.lookupFor(node)

    if (lookup) this.lookups.push(lookup)

    this.visitChildNodes(node)
  }

  private lookupFor(node: PrismNodes.CallNode): Lookup | null {
    const presence = PRESENCE_LOOKUPS.includes(node.name)

    if (!presence && !VALUE_LOOKUPS.includes(node.name)) return null
    if (!this.isLocalAssignsRead(node.receiver)) return null

    const argumentNodes = node.arguments_?.arguments_ ?? []
    const [first, ...rest] = argumentNodes

    if (!isPrismNodeType(first, "SymbolNode")) return null
    if (rest.length > 0 && node.name !== "fetch") return null
    if (rest.length > 1) return null

    const name = first.unescaped?.value

    if (!name) return null

    return {
      name,
      method: node.name,
      presence,
      extraArguments: rest.length > 0,
      startOffset: node.location.startOffset,
      length: node.location.length,
    }
  }

  private isLocalAssignsRead(node: PrismNodes.Node | null): boolean {
    if (!isPrismNodeType(node, "CallNode")) return false
    if (node.receiver !== null) return false

    return node.name === LOCAL_ASSIGNS
  }
}

export class ActionViewNoRedundantLocalAssignsRule extends ParserRule {
  static ruleName = "actionview-no-redundant-local-assigns"
  static introducedIn = this.version("unreleased")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: {
        cli: "error",
        editor: "info",
      },
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      strict_locals: true,
      prism_program: true,
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    if (isPartialFile(context?.fileName) === false) return []

    const document = result.value
    const source = document.source
    const program = document.prismNode

    if (!source || !program) return []

    const declaration = this.declarationFor(document)

    if (!declaration.hasDeclaration) return []

    const collector = new LocalAssignsLookupCollector()

    collector.visit(program)

    return collector.lookups.flatMap(lookup => {
      const message = this.messageFor(lookup, declaration)

      if (!message) return []

      const location = locationFromByteOffset(source, lookup.startOffset, lookup.length)

      return [this.createOffense(message, location, undefined, undefined, ["unnecessary"])]
    })
  }

  private declarationFor(document: DocumentNode): StrictLocalsCollector {
    const collector = new StrictLocalsCollector()

    collector.visit(document)

    return collector
  }

  private messageFor(lookup: Lookup, declaration: StrictLocalsCollector): string | null {
    const local = declaration.locals.get(lookup.name)
    const source = this.lookupSource(lookup)

    if (!local) {
      if (declaration.hasKeywordRest) return null

      return `\`${lookup.name}\` is not declared in the \`locals:\` declaration, so Rails raises if a caller passes it and \`${source}\` can never find it. Declare \`${lookup.name}:\` in the declaration, or remove the lookup.`
    }

    if (!local.required) return null

    if (lookup.presence) {
      return `Strict local \`${lookup.name}\` is required, so \`${source}\` is always \`true\`. Remove the condition, or give \`${lookup.name}\` a default value to make it optional.`
    }

    return `Strict local \`${lookup.name}\` is already a local variable in this partial, so \`${source}\` reads back a value that is already in scope. Use \`${lookup.name}\` instead.`
  }

  private lookupSource(lookup: Lookup): string {
    const argument = lookup.extraArguments ? `:${lookup.name}, ...` : `:${lookup.name}`

    return lookup.method === "[]" ? `${LOCAL_ASSIGNS}[${argument}]` : `${LOCAL_ASSIGNS}.${lookup.method}(${argument})`
  }
}

File: rust/herb-linter/src/rules/actionview_no_redundant_local_assigns.rs
use std::collections::HashMap;

use crate::offense::UnboundOffense;
use crate::rule::{LintContext, ParserRule, Rule};
use crate::utils::file_utils::is_partial_file;
use crate::utils::prism_utils::walk_prism;
use crate::utils::source_slice::location_from_offset;

use herb::nodes::{AnyNode, ERBStrictLocalsNode};
use herb::prism::PrismNode;
use herb::ParseResult;
use herb::Visitor;
use herb_config::{Severity, SeverityConfig};

const KEYWORD_KIND: &str = "keyword";
const KEYWORD_REST_KIND: &str = "keyword_rest";
const LOCAL_ASSIGNS: &str = "local_assigns";

const VALUE_LOOKUPS: &[&str] = &["[]", "fetch", "dig"];
const PRESENCE_LOOKUPS: &[&str] = &["key?", "has_key?", "include?", "member?"];

pub struct ActionViewNoRedundantLocalAssignsRule;

impl Rule for ActionViewNoRedundantLocalAssignsRule {
  fn name(&self) -> &'static str {
    "actionview-no-redundant-local-assigns"
  }

  fn introduced_in(&self) -> Option<&'static str> {
    Some("unreleased")
  }

  fn default_severity(&self) -> SeverityConfig {
    SeverityConfig::PerMode {
      cli: Severity::Error,
      editor: Severity::Info,
    }
  }

  fn parser_options(&self) -> herb::ParserOptions {
    herb::ParserOptions {
      strict_locals: true,
      prism_program: true,
      ..crate::rule::default_linter_parser_options()
    }
  }
}

struct Lookup {
  name: String,
  method: String,
  presence: bool,
  extra_arguments: bool,
  start_offset: usize,
  end_offset: usize,
}

impl Lookup {
  fn source(&self) -> String {
    let argument = if self.extra_arguments {
      format!(":{}, ...", self.name)
    } else {
      format!(":{}", self.name)
    };

    if self.method == "[]" {
      format!("{LOCAL_ASSIGNS}[{argument}]")
    } else {
      format!("{LOCAL_ASSIGNS}.{}({argument})", self.method)
    }
  }
}

#[derive(Default)]
struct StrictLocalsCollector {
  has_declaration: bool,
  has_keyword_rest: bool,
  locals: HashMap<String, bool>,
}

impl Visitor for StrictLocalsCollector {
  fn visit_erb_strict_locals_node(&mut self, node: &ERBStrictLocalsNode) {
    self.has_declaration = true;

    for local in &node.locals {
      let local = match local {
        AnyNode::RubyParameterNode(local) => local,
        _ => continue,
      };

      if local.kind == KEYWORD_REST_KIND {
        self.has_keyword_rest = true;
        continue;
      }

      if local.kind != KEYWORD_KIND {
        continue;
      }

      if let Some(name) = local.name.as_ref() {
        self.locals.insert(name.value.clone(), local.required);
      }
    }

    self.walk_erb_strict_locals_node(node);
  }
}

fn is_local_assigns_read(node: Option<&PrismNode>) -> bool {
  node.is_some_and(|node| node.is("CallNode") && node.receiver().is_none() && node.name.as_deref() == Some(LOCAL_ASSIGNS))
}

fn lookup_for(node: &PrismNode) -> Option<Lookup> {
  let method = node.name.as_deref()?;
  let presence = PRESENCE_LOOKUPS.contains(&method);

  if !presence && !VALUE_LOOKUPS.contains(&method) {
    return None;
  }

  if !is_local_assigns_read(node.receiver()) {
    return None;
  }

  let arguments = node.field_one("arguments").map(|arguments| arguments.field("arguments")).unwrap_or(&[]);
  let first = arguments.first()?;

  if !first.is("SymbolNode") {
    return None;
  }

  let rest = arguments.len() - 1;

  if (rest > 0 && method != "fetch") || rest > 1 {
    return None;
  }

  Some(Lookup {
    name: first.unescaped.clone()?,
    method: method.to_string(),
    presence,
    extra_arguments: rest > 0,
    start_offset: node.start_offset,
    end_offset: node.end_offset,
  })
}

fn message_for(lookup: &Lookup, declaration: &StrictLocalsCollector) -> Option<String> {
  let source = lookup.source();

  let required = match declaration.locals.get(&lookup.name) {
    Some(required) => *required,
    None => {
      if declaration.has_keyword_rest {
        return None;
      }

      return Some(format!(
        "`{}` is not declared in the `locals:` declaration, so Rails raises if a caller passes it and `{source}` can never find it. Declare `{}:` in the declaration, or remove the lookup.",
        lookup.name, lookup.name
      ));
    }
  };

  if !required {
    return None;
  }

  if lookup.presence {
    return Some(format!(
      "Strict local `{}` is required, so `{source}` is always `true`. Remove the condition, or give `{}` a default value to make it optional.",
      lookup.name, lookup.name
    ));
  }

  Some(format!(
    "Strict local `{}` is already a local variable in this partial, so `{source}` reads back a value that is already in scope. Use `{}` instead.",
    lookup.name, lookup.name
  ))
}

impl ParserRule for ActionViewNoRedundantLocalAssignsRule {
  fn check(&self, result: &ParseResult, context: &LintContext) -> Vec<UnboundOffense> {
    if let Some(file_name) = context.file_name.as_deref() {
      if !is_partial_file(file_name) {
        return Vec::new();
      }
    }

    let source = if context.source.is_empty() {
      result.source.clone()
    } else {
      context.source.clone()
    };

    let program = match result.value.prism() {
      Some(program) if !source.is_empty() => program,
      _ => return Vec::new(),
    };

    let mut declaration = StrictLocalsCollector::default();
    declaration.visit_document_node(&result.value);

    if !declaration.has_declaration {
      return Vec::new();
    }

    let mut offenses = Vec::new();

    walk_prism(program, &mut |node| {
      if node.is("CallNode") {
        if let Some(lookup) = lookup_for(node) {
          if let Some(message) = message_for(&lookup, &declaration) {
            offenses.push(UnboundOffense::with_tags(
              self.name(),
              message,
              location_from_offset(&source, lookup.start_offset, lookup.end_offset),
              vec!["unnecessary".to_string()],
            ));
          }
        }
      }

      true
    });

    offenses
  }
}
